### PR TITLE
Retry non-permanent HTTP failures

### PR DESCRIPTION
**Pull request by agoddard**

### CODE REVIEW

File: lib/jackal-kitchen/adjudicate.rb [hunk 1]
```diff
@@ -29,13 +29,13 @@ def execute(msg)
...
 
-          reasons = populate_reasons_for_failure(payload)
-          verdict = reasons.all? { |k, v| v.values.flatten.empty? }
+            reasons = populate_reasons_for_failure(payload)
+            verdict = reasons.all? { |k, v| v.values.flatten.empty? }
 
-          payload.set(:data, :kitchen, :judge, :decision, verdict)
-          job_completed(:kitchen, payload, msg)
+            payload.set(:data, :kitchen, :judge, :decision, verdict)
+            job_completed(:kitchen, payload, msg)
```
Comment (chrisroberts) [inline lib/jackal-kitchen/adjudicate.rb]: This is nested at the wrong depth

File: lib/jackal-kitchen/tester.rb [hunk 2]
```diff
@@ -22,7 +22,7 @@ def setup(*_)
       def valid?(msg)
         super do |payload|
           payload.get(:data, :code_fetcher, :asset) &&
-            !payload.get(:data, :kitchen, :test_output)
+            (payload.get(:data, :kitchen, :test_output).nil? || payload.get(:data, :kitchen, :retry))
```
Comment (chrisroberts) [inline lib/jackal-kitchen/tester.rb]: Keep `valid?` check same as previous (don't include retry check)

File: lib/jackal-kitchen/tester.rb [hunk 3]
```diff
@@ -78,6 +78,18 @@ def execute(msg)
...
+            h[:run_status][:http_failure][:permanent] == false
+          end
+          retry_count = payload.fetch(:data, :kitchen, :retry_count, 0)
+          retry_count += 1
+
+          payload.set(:data, :kitchen, :retry_count, retry_count) if failures
+
+          cond = failures && retry_count <= app_config.fetch(:kitchen, :config, :retries, 0)
+          payload.set(:data, :kitchen, :retry, cond)
+
```
Comment (chrisroberts) [inline lib/jackal-kitchen/tester.rb:94]: If retries are still allowed, delete `test_output` from payload

File: lib/jackal-kitchen/tester.rb [hunk 4]
```diff
@@ -102,8 +114,8 @@ def write_netrc
         begin
           gh_token = config.fetch(:github, :access_token,
                                   app_config.get(:github, :access_token))
-          git_host = config.fetch(:github, :uri,
-                                  app_config.get(:github, :uri))
+          git_host = config.fetch(config.fetch(:github, :uri,
```
Comment (chrisroberts) [inline lib/jackal-kitchen/tester.rb]: Remove outer most `config.fetch` and move default string in a level:

``` ruby
config.fetch(:github, :uri, app_config.fetch(:github, :uri, 'github.com'))
```

File: lib/jackal-kitchen/tester.rb (hunk 3 above)
Comment (agoddard) [inline lib/jackal-kitchen/tester.rb:94]: @chrisroberts ah ok cool, thanks. For some reason I was imagining that deleting from the payload was less advisable, but if it's a pattern, shall rock it.

File: lib/jackal-kitchen/tester.rb (hunk 4 above)
Comment (agoddard) [inline lib/jackal-kitchen/tester.rb]: as an interesting and related fact, the more you're nested in a .fetch method, the slower spacetime appears to the observer

Comment (chrisroberts) [inline lib/jackal-kitchen/tester.rb]: :space_invader:

File: lib/jackal-kitchen/adjudicate.rb (hunk 1 above)
Comment (ghost) [inline lib/jackal-kitchen/adjudicate.rb]: nice catch!

File: lib/jackal-kitchen/tester.rb (hunk 4 above)
Comment (ghost) [inline lib/jackal-kitchen/tester.rb]: hahaha

**Comment by ghost** [comment]
looks good;  \o/

**Pull request merged**